spec.bs

```diff
@@ -2236,6 +2248,38 @@ To <dfn>encode trusted signals keys</dfn> given an [=ordered set=] of [=strings=
 
 </div>
 
+A <dfn>bidding signals per interest group data</dfn> is a [=struct=] with the following
+[=struct/items=]:
+
+<dl dfn-for="bidding signals per interest group data">
+  : <dfn>updateIfOlderThanMs</dfn>
+  :: Null or a {{double}}. If non-null, it is the [=duration=] in milliseconds that specifies that
+    the [=interest group=] can be [updated](#interest-group-updates) if it hasn't been joined or

```

Comment:
Can you link to an algorithm that runs instead of just a header? I assume https://wicg.github.io/turtledove/#interest-group-update. Now that I think about it, where in this spec do we actually react to a timer and invoke the update algorithm? Sorry if I'm missing something.

Reply:
I'm not sure what you mean by "link to [...] a header"? The word "updated" here links to the section on interest group updating, which has algorithms.

This PR conditionally updates [=interest group/next update after=] for certain interest groups -- then, at this point [0], interest groups whose [=interest group/next update after=] is before the [=current wall time=] get updated.

(Note that I just noticed we shouldn't update interest groups that have an [=interest group/update url=] of null, so I fixed that in this PR.)

And yes, we always attempt an update for eligible interest groups, even though the update attempt might fail, with the exception that if we navigate away before the auction completes, the update attempt won't happen. We do continue trying to update for up to 10 minutes after navigating away (I don't think we put the give up after 10 minutes part in the spec...maybe that should be spec'd, though probably in another PR?)

[0] https://github.com/WICG/turtledove/blob/4de0e142338812c106c8e24d18b637166621e4cf/spec.bs#L4063

Reply:
> I'm not sure what you mean by "link to [...] a header"? The word "updated" here links to the section on interest group updating, which has algorithms.

I mean link to a specific algorithm, presumably https://wicg.github.io/turtledove/#interest-group-update.

> This PR conditionally updates [=interest group/next update after=] for certain interest groups -- then, at this point [0], interest groups whose [=interest group/next update after=] is before the [=current wall time=] get updated.

This part I'm a little unclear on. What mechanism do we use to actually trigger the update algorithm? I expected to see some timer set up with a callback pointing at https://wicg.github.io/turtledove/#interest-group-update that gets fired once the timer is up (based on `next update after` or something). Am I missing it?

> (I don't think we put the give up after 10 minutes part in the spec...maybe that should be spec'd, though probably in another PR?)

Yeah spec'ing this seems important, or giving it a mention at the very least. A separate PR sounds good.

Reply:
> I mean link to a specific algorithm, presumably https://wicg.github.io/turtledove/#interest-group-update.

Done.

> This part I'm a little unclear on. What mechanism do we use to actually trigger the update algorithm? I expected to see some timer set up with a callback pointing at https://wicg.github.io/turtledove/#interest-group-update that gets fired once the timer is up (based on next update after or something). Am I missing it?

There is no timer. Updates are triggered to run only by manual calls to `navigator.updateAdInterestGroups()` and post-auction updates (interest groups of buyers in the auction may update). In both cases, interest groups are only updated if the [=interest group/next update after=] is before now.

> Yeah spec'ing this seems important, or giving it a mention at the very least. A separate PR sounds good.

I threw in a quick note in this PR (both in the updates section, and the privacy considerations section) -- I think that should be sufficient, as I don't want to overspecify. Other browsers may be comfortable with different limits here, especially if they can implement some form of IP privacy.